Tests/SwiftFusionTests/Inference/MetropolisTests.swift

```diff
@@ -0,0 +1,287 @@
+    var values = graph.variablesZero
+
+    // Get some Metropolis samples.
+    let sampleCount = 5000
+    var samples: [Pose2] = []
+    samples.reserveCapacity(sampleCount)
+    for _ in 0..<sampleCount {
+      let maybeUpdate = graph.metropolisStep(
+        startState: values,
+        proposal: graph.exampleProposal,
+        using: &generator
+      )
+      if let update = maybeUpdate {
+        values.update(update)
+      }
+      samples.append(values[0])
+    }
+
+    // Calculate sample mean and covariance, in tangent space, as a simple test that the
+    // distribution is what we expect.
+    func mean(_ a: [Double]) -> Double {
+      return a.reduce(0, +) / Double(a.count)
+    }
+    func covariance(_ a: [Double], _ b: [Double]) -> Double {
+      precondition(a.count == b.count)
+      return mean(zip(a, b).map(*)) - mean(a) * mean(b)
+    }
+    let tangentSpaceSamples = samples.map { Pose2().localCoordinate($0) }
+    let xs = tangentSpaceSamples.map { $0.x }
+    let ys = tangentSpaceSamples.map { $0.y }
+    let zs = tangentSpaceSamples.map { $0.z }
+    XCTAssertEqual(mean(xs), 0, accuracy: 0.1)
+    XCTAssertEqual(mean(ys), 0, accuracy: 0.1)
+    XCTAssertEqual(mean(zs), 0, accuracy: 0.1)
+    XCTAssertEqual(covariance(xs, xs), 0.5, accuracy: 0.1)
+    XCTAssertEqual(covariance(ys, ys), 0.5, accuracy: 0.1)
+    XCTAssertEqual(covariance(zs, zs), 0.5, accuracy: 0.1)
+    XCTAssertEqual(covariance(xs, ys), 0.0, accuracy: 0.1)
+    XCTAssertEqual(covariance(xs, ys), 0.0, accuracy: 0.1)
+    XCTAssertEqual(covariance(ys, zs), 0.0, accuracy: 0.1)
+  }
+
+  /// Test the Metropolis algorithm on a simple Pose2SLAM problem.
+  func testMetropolisPose2SLAM() {
+    // Seeded RNG for reproducible results.
+    var generator = ThreefryRandomNumberGenerator(seed: [42])
+
+    // Construct a graph.
+    var graph = PoseSLAMFactorGraphForSmallProposals<Pose2>(variableCount: 5)
+    graph.add(0, prior: Pose2(0, 0, 0))
+    graph.add(1, 0, between: Pose2(2.0, 0.0, .pi / 2))
+    graph.add(2, 1, between: Pose2(2.0, 0.0, .pi / 2))
+    graph.add(3, 2, between: Pose2(2.0, 0.0, .pi / 2))
+    graph.add(4, 3, between: Pose2(2.0, 0.0, .pi / 2))
+
+    // Initial estimates for the poses.
+    var values = graph.variablesZero
+
+    // Mix it a bit.
+    let sampleCount = 5000
+    var acceptanceCount = 0
+    for _ in 0..<sampleCount {
+      let maybeUpdate = graph.metropolisStep(
+        startState: values,
+        proposal: graph.exampleProposal,
+        using: &generator
+      )
+      if let update = maybeUpdate {
+        values.update(update)
+        acceptanceCount += 1
+      }
+    }
+
+    // Metropolis with this proposal is terrible for this problem, so we're not getting anywhere
+    // near the solution. So just assert that there weren't 0 or 100% acceptances so that this
+    // test tests that Metropolis did some nontrivial work at least.
+    XCTAssertGreaterThan(acceptanceCount, 0)
```

Comment:
Maybe a good test is to start from the ground truth, and see if it works? I think metropolis with this *should* work, as this graph is pretty linear (like a Markov Chain)

Reply:
Good idea! I'll try that when I have some time.